### PR TITLE
Refcoder CFrame Special Cases & Fix

### PR DESCRIPTION
**Pull request by ArvidSilverlock**
Implemented [CFrame special cases](https://dom.rojo.space/binary.html#cframe) to reduce pack sizes, done via adding an extra byte to the packed value, each representing a CFrame's rotation (zero being a non-special case). Thus allowing CFrames with certain rotations to be stored the almost the same as a Vector3.

**Some other minor changes**
- Fixed error when decoding CFrames, `Invalid number of arguments: 13`, caused by passing the pointer returned by `string.unpack` into `CFrame.new`. Done by assigning a variable to each component.
- Added a guard clause into `getRefId` rather than a generic if statement (probably should revert, feels less readable)
- Changed some string interpolation quotes out for double quotes (no string interpolation was being done).

**Benchmarks for encoding/decoding CFrames**
image

**Test Script**
Ttests the special cases then tests random cases
```lua
local Refcoder = require(game.ReplicatedStorage.Modules.Refcoder)

local RAD_90 = math.rad(90)
local RAD_180 = math.rad(180)

local rotations = {
	CFrame.Angles(0, 0, 0),			CFrame.Angles(0, RAD_180, 0),
	CFrame.Angles(RAD_90, 0, 0),		CFrame.Angles(-RAD_90, -RAD_180, 0),
	CFrame.Angles(0, RAD_180, RAD_180),	CFrame.Angles(0, 0, RAD_180),
	CFrame.Angles(-RAD_90, 0, 0),		CFrame.Angles(RAD_90, RAD_180, 0),
	CFrame.Angles(0, RAD_180, RAD_90),	CFrame.Angles(0, 0, -RAD_90),
	CFrame.Angles(0, RAD_90, RAD_90),	CFrame.Angles(0, -RAD_90, -RAD_90),
	CFrame.Angles(0, 0, RAD_90),		CFrame.Angles(0, -RAD_180, -RAD_90),
	CFrame.Angles(0, -RAD_90, RAD_90),	CFrame.Angles(0, RAD_90, -RAD_90),
	CFrame.Angles(-RAD_90, -RAD_90, 0),	CFrame.Angles(RAD_90, RAD_90, 0),
	CFrame.Angles(0, -RAD_90, 0),		CFrame.Angles(0, RAD_90, 0),
	CFrame.Angles(RAD_90, -RAD_90, 0),	CFrame.Angles(-RAD_90, RAD_90, 0),
	CFrame.Angles(0, RAD_90, RAD_180),	CFrame.Angles(0, -RAD_90, RAD_180),
}

local checksPassed = 0
local checkCount = 100000

local encodeTime = 0
local decodeTime = 0

local function checkCFrame(cframe: CFrame)
	local startTime = os.clock()
	local encoded = Refcoder.encode(cframe)
	local middleTime = os.clock()
	local decoded = Refcoder.decode(encoded)
	local endTime = os.clock()

	if cframe == decoded then
		encodeTime += middleTime - startTime
		decodeTime += endTime - middleTime
		checksPassed += 1
	else
		warn(`Expected {cframe}, got {decoded}`)
	end
end

for _, rotation in rotations do
	local cframe = CFrame.new(math.random() * 25, math.random() * 25, math.random() * 25) * rotation
	checkCFrame(cframe)
end

for _ = 1, checkCount - #rotations do
	local cframe = CFrame.new(math.random() * 25, math.random() * 25, math.random() * 25)
		* CFrame.Angles(math.random() * 25, math.random() * 25, math.random() * 25)
	checkCFrame(cframe)
end

print(`Passed {checksPassed}/{checkCount} checks.`)
print(`Took {encodeTime / checksPassed * 1000000}μs to encode on average.`)
print(`Took {decodeTime / checksPassed * 1000000}μs to decode on average.`)
```

### CODE REVIEW

**Comment by Hexcede** [comment]
The lack of a guard clause when initializing `refId` is actually a conscious style choice. I use guard clauses in almost all other cases, but when initializing values from tables like that, I tend to prefer not using a guard clause and instead initializing the value in an if statement, since a lot of the time the value gets immediately used (it's just that in this particular case, it doesn't get used).

I believe in the `metadata-refactor` branch `refId` does get used after the if statement now (although I didn't check so I might be remembering wrong).

File: src/init.luau [hunk 1]
```diff
@@ -79,31 +98,32 @@ end
...
+	local encodingId = state.EncodingsToRefs[packed]
+	if encodingId then
+		-- Replace with a copy_of for the encoded data
+		packed = encodeValue(encodeRefId(encodingId), TYPE_COPY_OF)
+	else
+		-- Store the ref ID of this data
+		state.EncodingsToRefs[packed] = refId
 	end
+	
+	state.Encodings[refId] = packed
```
Comment (Hexcede) [inline src/init.luau]: Should revert the changes here back to the original, since it's a conscious style choice.

File: src/init.luau [hunk 2]
```diff
@@ -54,6 +54,25 @@ local function decodeRefTable(data: string, readStart: number?)
...
+	CFrame.Angles(0, RAD_180, RAD_90);	CFrame.Angles(0, 0, -RAD_90);
+	CFrame.Angles(0, RAD_90, RAD_90);	CFrame.Angles(0, -RAD_90, -RAD_90);
+	CFrame.Angles(0, 0, RAD_90);		CFrame.Angles(0, -RAD_180, -RAD_90);
+	CFrame.Angles(0, -RAD_90, RAD_90);	CFrame.Angles(0, RAD_90, -RAD_90);
+	CFrame.Angles(-RAD_90, -RAD_90, 0);	CFrame.Angles(RAD_90, RAD_90, 0);
+	CFrame.Angles(0, -RAD_90, 0);		CFrame.Angles(0, RAD_90, 0);
+	CFrame.Angles(RAD_90, -RAD_90, 0);	CFrame.Angles(-RAD_90, RAD_90, 0);
+	CFrame.Angles(0, RAD_90, RAD_180);	CFrame.Angles(0, -RAD_90, RAD_180);
+}
+
```
Comment (Hexcede) [inline src/init.luau:75]: Should consider generating this table automatically.

File: src/init.luau [hunk 3]
```diff
@@ -125,7 +145,9 @@ function encode(value: any, state: EncodeState, isStatic: boolean?)
 	elseif typeof(value) == "Vector3" then
 		return string.pack("<nnn", value.X, value.Y, value.Z), typeIndex
 	elseif typeof(value) == "CFrame" then
-		return string.pack(`<{string.rep("nnn", 4)}`, value:GetComponents()), typeIndex
+		local caseId = table.find(CFRAME_SPECIAL_CASES, value.Rotation)
+		local packFormat = string.rep("nnn", if caseId == nil then 4 else 1)
+		return string.pack(`<B<{packFormat}`, caseId or 0, value:GetComponents()), typeIndex
```
Comment (Hexcede) [inline src/init.luau]: Second `<` should be removed since it's unnecessary.

File: src/init.luau [hunk 4]
```diff
@@ -125,7 +145,9 @@ function encode(value: any, state: EncodeState, isStatic: boolean?)
 	elseif typeof(value) == "Vector3" then
 		return string.pack("<nnn", value.X, value.Y, value.Z), typeIndex
 	elseif typeof(value) == "CFrame" then
-		return string.pack(`<{string.rep("nnn", 4)}`, value:GetComponents()), typeIndex
+		local caseId = table.find(CFRAME_SPECIAL_CASES, value.Rotation)
```
Comment (Hexcede) [inline src/init.luau:147]: Right now `caseId` doesn't really do anything beneficial, since below the CFrame's full components list is still saved anyways, so I am not entirely sure I understand what the point of it is. I am making a guess that you originally meant to conditionally save the components.

This should save only the `caseId` byte if the CFrame is in the special cases list and three numbers for the position. If it isn't (`caseId` 0) the entire CFrame's components should get saved with the byte as just `0`.

This seems to already be accounted for in the code below.

File: src/init.luau [hunk 5]
```diff
@@ -165,13 +189,20 @@ local function decode(refId: number, encoded: string, state: DecodeState)
 	elseif valueType == "Vector3" then
 		return Vector3.new(string.unpack("<nnn", value))
 	elseif valueType == "CFrame" then
-		return CFrame.new(string.unpack(`<{string.rep("nnn", 4)}`, value))
+		local cframeCase, pointer = string.unpack("<B", value)
+		if cframeCase ~= 0 then
```
Comment (Hexcede) [inline src/init.luau]: Should flip the branches in this if statement so the condition is positive.

File: src/init.luau [hunk 6]
```diff
@@ -165,13 +189,20 @@ local function decode(refId: number, encoded: string, state: DecodeState)
...
+			local x, y, z = string.unpack("<nnn", value, pointer)
+			return CFrame.new(x, y, z) * CFRAME_SPECIAL_CASES[cframeCase]
```
Comment (Hexcede) [inline src/init.luau]: Instead of CFrame multiplication, let's use this instead:
```lua
local position = Vector3.new(string.unpack("<nnn", value, pointer))
local orientation = CFRAME_SPECIAL_CASES[cframeCase]
return CFrame.fromMatrix(position, orientation.XVector, orientation.YVector, orientation.ZVector)
```

File: src/init.luau [hunk 7]
```diff
@@ -165,13 +189,20 @@ local function decode(refId: number, encoded: string, state: DecodeState)
...
+		else
+			local x, y, z, R00, R01, R02, R10, R11, R12, R20, R21, R22 = string.unpack(`<{string.rep("nnn", 4)}`, value, pointer)
```
Comment (Hexcede) [inline src/init.luau]: Maybe consider using `CFrame.fromMatrix` here as well, and loading `XVector`, `YVector`, and `ZVector` separately instead? I think this would be cleaner and more readable. I believe this would require serializing position, `XVector`, `YVector`, and `ZVector` above since I don't think the components are in the same order (pretty sure R00 R10 R20 makes up XVector instead of R00 R01 R02 but I might be wrong, should definitely check this)

**Review by Hexcede — CHANGES_REQUESTED** [review]

File: src/init.luau (hunk 2 above)
Comment (ArvidSilverlock) [reply src/init.luau:75]: This table is not all of the 90° snapped rotations, it's specifically the ones that `BasePart` will set their rotations to, for example, setting a parts `CFrame` to '180, 0, 0' actually sets it to `0, 180, 180`.

I'm unsure of the rhyme or reason behind these values, and thus cannot generate it automatically.

File: src/init.luau (hunk 4 above)
Comment (ArvidSilverlock) [reply src/init.luau:147]: The `packFormat` variable (defined as `string.rep("nnn", if caseId == nil then 4 else 1)`) would be "nnnnnnnnnnnn" if `caseId` is nil (meaning it is not a special CFrame case), whereas it would be "nnn" if `caseId` *isn't* nil, which means it would only store the position, and not the rotation matrix.

The reason I still store `caseId`, even if the full list of components is still present, is because I am unsure on how I would identify that it is a normal CFrame with all components listed when decoding.

File: src/init.luau (hunk 7 above)
Comment (ArvidSilverlock) [reply src/init.luau]: I'm not too sure how I would achieve this whilst retaining the same performance, as if i'm not wrong, you want construct 4 `Vector3`s (whilst unpacking 4 separate times?)


According to the [CFrame docs](https://create.roblox.com/docs/reference/engine/datatypes/CFrame), R00, R10 and R20 make up to RightVector, but the XVector is made up of R00, R01 and R02.
image
